### ISSUE TITLE
gnome-shell crash report's var_log_messages file is useless

### ISSUE THREAD

**Issue by mchalupa** [abrt/abrt#1030]
The usual contents is something like:
```
User Logs:
-- Logs begin at Di 2015-08-11 21:31:29 CEST, end at Do 2015-08-13 12:09:15 CEST. --
Aug 13 12:09:13 localhost.localdomain audit[1930]: ANOM_ABEND auid=1000 uid=1000 gid=1000 ses=1 subj=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 pid=1930 comm="gnome-shell" exe="/usr/bin/gnome-shell" sig=5
```

Which won't tell developers much really. The problem is that gnome-shell logs are going through gnome-session, so the relevant logs are from gnome-session, not gnome-shell. Locally, I added this script to /etc/libreport/events.d that fixes it for me:

```shell
# save gnome-session logs with gnome-shell crash

# use regexp instead of component=gnome-shell,
# because sometimes the component is not set
EVENT=post-create executable~=/gnome-shell$
        # save logs from gnome-session since it logs everything from gnome-shell
        journalctl --since=-3m _COMM=gnome-session > $DUMP_DIR/gnome_session_log
        true
```

**Comment by jfilak** [comment]
Brilliant job! Thank you!  Would it be possible to ship your dump-gnome-shell-logs script in gnome-shell package?

**Comment by owtaylor** [comment]
Hmm, Marek - I think a better approach would be to make gnome-shell redirect it's log to systemd with the correct unit if !isatty(1).

This could be done two ways:

 - Using g_log_set_handler()
 - Using sd_journal_stream_fd() [as we do for launched apps, see shell-app.c]

Probably the first is better, since the logging can be done with the correct priorities, JS backtraces can be logged as a single log message instead of one-per-line, etc. And we're pretty good about getting all messages through the log system already. Do you want to try coming up with a patch for that?

**Comment by jfilak** [comment]
@owtaylor I think gnome-shell writes its log to systemd-journald but ABRT fails to find those messages because of this https://github.com/abrt/abrt/blob/master/src/plugins/ccpp_event.conf#L32 We plan to develop a far better systemd-journald logs dumper (#864 @mhabrnal will work on it once he finishes the xorg dumper).

Anyway, writing JS backtraces to systemd-jornald is what we really need.  I asked GNOME developers to write it to a file but that's an insecure approach.

**Comment by owtaylor** [comment]
Yeah - the _COMM="$base_executable" is what is not working for GNOME Shell - GNOME Shell just logs to stderr, and that gets resent to the log by gnome-session with _COMM=gnome-session. What I'm saying is that I'd rather fix that in gnome-shell, rather then ship a workaround where we send everything logged by gnome-session to 

JS backtraces is a slightly different topic - there's a couple of different cases:

 - An uncaught exception happens without a crash. This isn't my *biggest* concern - it might cause misbehavior, but hopefully the user doesn't lose data. But picking them out of the log and reporting them would certainly give us more data to work with to fix bugs. (And also could work with retrace.fedoraproject.org)

 - A backtrace is logged then subsequently gnome-shell crashes - this would be picked up if we fixed including log entries with C crashes as discussed above.

 - A crash occurs *without* an exception in Javascript, but we'd really like to have the JS stack with the report. While in some theoretical world it would be possible to write gdb code to do a post-mortem JS backtrace, in practice that would prohibitive. The place where this is easier to fix is when the crash occurs via g_assert() or g_error() - then we can just have the error handle call gjs_dumpstack() and have some reasonable hope that it will work. The place where it's harder to handle is a segfault or non-glib assertion failure. Calling gjs_dumpstack() from a SEGV handler is scary - I suppose the worst that can happen is a segfault :-) but in the process we might lose a useful C backtrace.

**Comment by sorki** [comment]
The best option for gnome-shell would be to allow post mortem analysis so it's possible to extract JS backtrace from coredumps.

**Comment by owtaylor** [comment]
Agreed - but my suspicion is that's a 1000-line gdb Python script that will break every time a new spidermonkey release happens. :-)

**Comment by mchalupa** [comment]
is there a way how to get JS backtrace as a single line?

**Comment by owtaylor** [comment]
I'm not really sure I understand the question - when GJS logs an exception, it logs it as a single call to g_logv() - the message may contain embedded newlines. systemd journal log messages can contain embedded newlines, so there shouldn't be a problem with propagating it like that.

**Comment by mchalupa** [comment]
with g_logv() is really no problem, but gnome-shell's log handler uses gjs_dumpstack() which prints to stderr https://git.gnome.org/browse/gnome-shell/tree/src/main.c#n322

**Comment by owtaylor** [comment]
Hmm, that's only hit for SHELL_DEBUG=backtrace-warnings, which is not set by default. I think it would be fine as a first pass to not worry about it and let that get logged as gnome-session. Beyond that:

 A) Nothing in gjs_dumpstack() uses private API - you can cut-and-paste it (or better gjs_context_print_stack_stderr) and modify as necessary.

 B) I think some function to get the stack a string could be added to a future version of GJS.

**Comment by mchalupa** [comment]
https://bugzilla.gnome.org/show_bug.cgi?id=757571
